design-proposals/klip-8-interactive-queries.md

```diff
@@ -131,13 +131,129 @@ Additionally:
+
+```sql
+-- Streaming query on a stream
+SELECT x, count(*) FROM stream GROUP BY x EMIT CHANGES;
+
+-- Streaming query on a table
+SELECT x, y FROM table WHERE x < 10 EMIT CHANGES;
+
+-- Point-in-time query on a table
+SELECT ROWKEY, count FROM table WHERE ROWKEY = 'key';
+
+-- Point-in-time query on a stream (semantics unclear, illustrative example only)
+SELECT * FROM stream WHERE column = 42;
+```
+
+---
+
+Finally, `EMIT CHANGES` may be preferrable over `EMIT STREAM` for KSQL specifically (for the reasons outlined above) although `EMIT STREAM` is still a suitable option.
+
+## Proposed behavioral changes
+
+As we add this fundamentally new capability to KSQL, it is worth considering any modifications we can make to KSQL's existing behavior to ensure query behavior is as intuitive as possible after the addition of PIT queries. There is one potential behavioral change that is worth considering in this respect: **streaming queries against a table should return the table's entire materialized state (possibly filtered by a `WHERE` clause) followed by changes.**
+
+Currently, a streaming query against a table in KSQL will only return rows from the table's changelog produced *after* the query begins (assuming a default start offset of `latest`).
+
+However, streaming queries returning a table's entire materialized state *then* followed by changes is more consistent with the behavior of PIT queries on a table. This behavior is also likely what users want when issuing a streaming query against a table, and it is currently not straightforward to achieve.
+
+To get KSQL's current behavior with a streaming query against a table, users would simply use an intuitive `WHERE` clause:
+
+```sql
+SELECT * FROM table WHERE ROWTIME >= NOW EMIT CHANGES;
+```
+
+## Limitations
+
+The following are proposed intentional limitations of the initial implementation of PIT queries:
+
+**PIT queries on streams probably do not need to be initially supported.** The semantics and use cases around these are unclear and require further analysis. However, the proposed syntax leaves open the possibility of supporting PIT queries on streams without further modifications.
+
+**CTAS/CSAS defined by PIT queries do not initially need to be supported.** For example, the following CTAS would create a table, `t0`, and initially populate it with the result of the `SELECT` query:
+
+```sql
+CREATE TABLE t0 AS SELECT x, count(*) FROM t1 GROUP BY x;
+```
+
+**PIT queries against tables created directly over a topic (leaf tables) do not need to be initially supported.** Leaf tables are not materialized until another query reads from them. As a result, a PIT query against such a table would (potentially) not return the expected result. Rather than leave open the possibility of confusing query results in this scenario, we should simply disallow PIT queries against leaf tables until we decide to support them.
```

Comment:
I can see why, because of LOE and timeline, we might not be able to also add this functionality. That's ok and makes sense.

That said, I'd imagine that the first attempts of many users would start with something simple: You have a Kafka topic, which you want to read into a table and then (PIT) query like an RDBMS table. So this is a "first 15 minutes" feature that we should add sooner rather than later. Similarly, the upcoming KConnect integration is another reason why we should support this, since I understand that in the KConnect scenario the user, too, would not be able to query a "leaf" table. TL;DR: I consider this feature table-stakes, but if we can't implement it right away, that's acceptable.

Reply:
> Should we 'pretend' that these queries are a PIT but actually scan the entire changelog every time? 

This would not be _pretending_ anything. Scanning the changelog up to the current offset _is_ a PIT query. It's just not as efficient as querying an already materialized state store.